pkg/playback/interactions.go

```diff
@@ -25,126 +21,101 @@ const (
 	incomingInteraction                        // eg: webhooks
 )
 
-// cassettePairs stores info on a single request + response interaction
-type cassettePair struct {
+// interaction stores info on a single request + response interaction
+// interactions are on the tape ready to be persisted, so Request/Response
+// are interface{} - already encoded by the serializer, ready to be persisted.
+type interaction struct {
 	// may have other fields like -- sequence number
 	Type     interactionType
-	Request  []byte
-	Response []byte
+	Request  httpRequest
+	Response httpResponse
```

Comment:
I want to argue for making these generic (i.e. `interface{}` since Go doesn't yet have generics).

Briefly the way I conceive of this module is like this:
- `interactionRecorder` takes `Stream<Out>`, `Stream<In>` and produces `Recording<(In, Out)>`.
- `interactionReplayer` takes `Comparator<Out>`, `Recording<(In, Out)>`, `Stream<Out>` and produces a `Stream<In>`

It looks like the direction in this PR is to specialize `In` to `httpRequest`, and `Out` to `httpResponse` (not exactly -- it's the other way around depending on "Type") and specialize "Recording" to yaml (you haven't specialized Comparator, though -- but if we should if we are specializing the other things). The advantage of this is that it removes some indirection and makes the code a bit easier to follow (for now), but I think architecturally we should try to enforce that the general logic of "replaying recorded events" is not coupled to the fact that those events happen to represent HTTP requests/responses (that logic should live entirely in the comparator), or that the serialization format happens to be yaml (that logic should live entirely in the serializer). This will help us out in the event we want to support e.g. multiple serialization formats, and will also help us if we need to include more types of "interactions" in a recording (not too likely, honestly, but conceivable -- timed out http requests? grpc? special stripe-cli directives?)

I don't consider this blocking at all, if you're not persuaded. The lack of generics unfortunately means this all has non-trivial cost in terms of indirection, but I wanted to share how I think about this code & the idea for your consideration.

Reply:
Also I recommend this talk if you haven't seen it, https://www.softwaretalks.io/v/483/constraints-liberate-liberties-constrain-runar-bjarnason -- it's heavily influenced how I think about avoiding specialized implementations.

Reply:
I agree with you and it's almost what I was trying to go towards;

- the refactoring towards the `InteractionRecorder` and `InteractionReplayer` taking a `serializer` is to allow deciding how we want to serialize things to disk (yaml in this PR). I might have miss some places but don't think there is any serialization leaks outside of the serializer.
Do you have any recommendations on how to make the serializer system more isolated?

- For the `httpRequest` / `httpResponse` being `interface{}` I kind of agree I supposed, I had `interface{}` originally but it was giving me some headaches, I can't recall the details - and at the same time I don't know if we'll _really_ have any different types of streams, since libraries communicate with the Stripe API by http and that's what we'll probably be dealing with for the next few years?

Reply:
No I took a second look and I think I was confused, reading a function call as an interface or something. What you have looks perfect. :shipit: